File: app/src/main/java/com/coinninja/coinkeeper/ui/transaction/SyncManagerViewNotifier.java
package com.coinninja.coinkeeper.ui.transaction;

import android.os.Handler;
import android.os.Looper;

import com.coinninja.coinkeeper.di.interfaces.CoinkeeperApplicationScope;
import com.coinninja.coinkeeper.ui.transaction.history.SyncManagerChangeObserver;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

@CoinkeeperApplicationScope
public class SyncManagerViewNotifier implements SyncManagerChangeObserver {
    private List<WeakReference<SyncManagerChangeObserver>> observers;

    private boolean isSyncing = false;

    @Inject
    SyncManagerViewNotifier() {
        observers = new ArrayList<>();
    }

    public void observeSyncManagerChange(SyncManagerChangeObserver observer) {
        observers.add(new WeakReference<>(observer));
    }

    public void onSyncStatusChanged() {
        for (WeakReference<SyncManagerChangeObserver> observerWeakReference : observers) {
            if (observerWeakReference.get() != null) {
                observerWeakReference.get().onSyncStatusChanged();
            }
        }
    }

    synchronized public boolean isSyncing() {
        return isSyncing;
    }

    synchronized public void setSyncing(boolean syncing) {
        isSyncing = syncing;
        Handler main = new Handler(Looper.getMainLooper());
        main.post(() -> {
            onSyncStatusChanged();
        });
    }
}

File: app/src/main/java/com/coinninja/coinkeeper/ui/transaction/SyncManagerViewNotifier.kt
package com.coinninja.coinkeeper.ui.transaction

import android.os.Handler
import android.os.Looper
import app.dropbit.annotations.Mockable
import com.coinninja.coinkeeper.di.interfaces.CoinkeeperApplicationScope
import com.coinninja.coinkeeper.ui.transaction.history.SyncManagerChangeObserver
import java.lang.ref.WeakReference
import java.util.*
import javax.inject.Inject

@Mockable
@CoinkeeperApplicationScope
class SyncManagerViewNotifier @Inject internal constructor() : SyncManagerChangeObserver {

    private val observers: MutableList<WeakReference<SyncManagerChangeObserver?>>

    @get:Synchronized
    var isSyncing = false
        @Synchronized set(syncing) {
            field = syncing
            val main = Handler(Looper.getMainLooper())
            main.post { this.onSyncStatusChanged() }
        }

    init {
        observers = ArrayList()
    }

    fun observeSyncManagerChange(observer: SyncManagerChangeObserver?) {
        observers.add(WeakReference(observer))
    }

    override fun onSyncStatusChanged() {
        for (observerWeakReference in observers) {
            observerWeakReference.get()?.onSyncStatusChanged()
        }
    }
}
